File: android/src/main/kotlin/com/technit/capacitor/conekta/ConektaTokenizerPlugin.kt
package com.technit.capacitor.conekta

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "ConektaTokenizer")
class ConektaTokenizerPlugin : Plugin() {

    private val implementation = ConektaTokenizer()

    override fun load() {
        implementation.setup(activity)
    }

    @PluginMethod
    fun setPublicKey(call: PluginCall) {
        val publicKey = call.getString("publicKey")
        if (publicKey == null) {
            call.reject("publicKey is required")
            return
        }
        implementation.setPublicKey(publicKey)
        call.resolve()
    }

    @PluginMethod
    fun createToken(call: PluginCall) {
        val name = call.getString("name")
        val cardNumber = call.getString("cardNumber")
        val expMonth = call.getString("expMonth")
        val expYear = call.getString("expYear")
        val cvc = call.getString("cvc")

        if (name == null || cardNumber == null || expMonth == null || expYear == null || cvc == null) {
            call.reject("All card fields are required: name, cardNumber, expMonth, expYear, cvc")
            return
        }

        implementation.createToken(
            name = name,
            cardNumber = cardNumber,
            expMonth = expMonth,
            expYear = expYear,
            cvc = cvc,
            onSuccess = { tokenObj ->
                val ret = JSObject()
                ret.put("token", JSObject(tokenObj.toString()))
                call.resolve(ret)
            },
            onError = { raw ->
                val message = when {
                    raw.has("message_to_purchaser") -> raw.optString("message_to_purchaser")
                    raw.has("message") -> raw.optString("message")
                    else -> "Token creation failed"
                }
                val code = if (raw.has("code")) raw.optString("code") else null
                val data = JSObject().apply {
                    put("conektaError", JSObject(raw.toString()))
                    if (raw.has("type")) put("type", raw.optString("type"))
                    if (raw.has("param")) put("param", raw.optString("param"))
                }
                call.reject(message, code, null, data)
            }
        )
    }
}

File: ios/Sources/ConektaTokenizerPlugin/ConektaTokenizerPlugin.swift
import Foundation
import Capacitor

@objc(ConektaTokenizerPlugin)
public class ConektaTokenizerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ConektaTokenizerPlugin"
    public let jsName = "ConektaTokenizer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setPublicKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "warmUp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isReady", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = ConektaTokenizer()

    override public func load() {
        implementation.setup()
    }

    @objc func setPublicKey(_ call: CAPPluginCall) {
        guard let publicKey = call.getString("publicKey") else {
            call.reject("publicKey is required")
            return
        }
        implementation.setPublicKey(publicKey)
        call.resolve()
    }

    @objc func isReady(_ call: CAPPluginCall) {
        call.resolve(["ready": implementation.isReady()])
    }

    @objc func warmUp(_ call: CAPPluginCall) {
        implementation.warmUp { result in
            switch result {
            case .success:
                call.resolve(["ready": true])
            case .failure(let error):
                self.rejectWithConektaError(call: call, error: error)
            }
        }
    }

    @objc func createToken(_ call: CAPPluginCall) {
        guard let name = call.getString("name"),
              let cardNumber = call.getString("cardNumber"),
              let expMonth = call.getString("expMonth"),
              let expYear = call.getString("expYear"),
              let cvc = call.getString("cvc") else {
            call.reject("All card fields are required: name, cardNumber, expMonth, expYear, cvc")
            return
        }

        implementation.createToken(
            name: name,
            cardNumber: cardNumber,
            expMonth: expMonth,
            expYear: expYear,
            cvc: cvc
        ) { result in
            switch result {
            case .success(let tokenObj):
                call.resolve(["token": tokenObj])
            case .failure(let error):
                self.rejectWithConektaError(call: call, error: error)
            }
        }
    }

    private func rejectWithConektaError(call: CAPPluginCall, error: Error) {
        if case let ConektaError.apiError(message, code, type, param, raw) = error {
            var data: [String: Any] = ["conektaError": raw]
            if let type = type { data["type"] = type }
            if let param = param { data["param"] = param }
            call.reject(message, code, error, data)
            return
        }
        if let conektaError = error as? ConektaError {
            call.reject(conektaError.errorDescription ?? "Conekta error", conektaError.code, error)
            return
        }
        call.reject(error.localizedDescription, nil, error)
    }
}
